_blog/2024-01-23-using-serde-rust.mdx

```diff
@@ -0,0 +1,341 @@
+---
+title: "Using Serde in Rust"
+description: This article talks about serde, a Rust serialization library and how you can use it in applications.
+author: josh
+tags: [rust, serde, serialization, guide]
+thumb: 
+cover: 
+date: '2024-01-23T14:30:00'
+---
+In this article we’ll be talking about Serde, how you can use it in your Rust application as well as some more advanced tips and tricks.
+
+## What is serde?
+
+The `serde` Rust crate is used to efficiently serialize and deserialize data in many formats. It does this by providing two traits you can use, aptly named `Deserialize` and `Serialize`. Being one of the most well-known crates in the ecosystem, it currently supports (de)serialization to over 20 types.
+
+To get started, you’ll want to install the crate into your Rust application:
+
+```bash
+cargo add serde
+```
+
+## Using serde
+
+### Deserializing and Serializing data
+
+The simple way to serialize and deserialize data is by adding the serde `derive` feature. This adds a macro that you can use to implement `Deserialize` and `Serialize` automatically - you can do this with the `--features` flag (`-F` for short):
+
+```bash
+cargo add serde -F derive
+```
+
+Then we can add a macro to any struct or enum that we want to implement `Deserialize` or `Serialize` for:
+
+```rust
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+struct MyStruct {
+	message: String,
+    // ... the rest of your fields
+}
+```
+
+This allows us to use any crate with `serde` support to convert between said formats. As an example, let’s use `serde-json` to convert to and from JSON format:
+
+```rust
+use serde_json::json;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+struct MyStruct {
+    message: String,
+}
+
+fn to_and_from_json() {
+    let json = json!({"message": "Hello world!"});
+    let my_struct: MyStruct = serde_json::from_str(&json).unwrap();
+
+    assert_eq!(my_struct, MyStruct { message: "Hello world!".to_string());
+
+    assert!(serde_json::to_string(my_struct).is_ok());
+}
+```
+
+If you’re interested in using `serde-json` for your Rust application, we have an article talking about JSON parsing libraries which you can check out [here.](https://www.shuttle.rs/blog/2024/01/18/parsing-json-rust)  
+
+We can also deserialize and serialize to/from many sources including from a file stream I/O, a JSON byte array and more!
+
+### Implementing Deserialize and Serialize manually
+
+As a side note, you can also implement `Deserialize` and `Serialize` manually! This is quite complicated, but for now we will stick with a simple implementation. Here is a simple implementation for serializing of the `i32` primitive type. 
```

Comment:
Is it really a side note if you're dedicating a whole paragraph to it ? I feel like it would make more sense to say that "in order to understand a bit more how it is working under the hood, we can implement `Serialize` and `Deserialize` ourselves".